storage/store.go

```diff
@@ -175,42 +193,27 @@ func (e *NotBootstrappedError) Error() string {
 	return "store has not been bootstrapped"
 }
 
-// storeRangeIterator is an implementation of rangeIterator which
-// cycles through a store's rangesByKey slice.
-type storeRangeIterator struct {
-	store     *Store
-	remaining int
-	index     int
-}
-
-func newStoreRangeIterator(store *Store) *storeRangeIterator {
-	r := &storeRangeIterator{
-		store: store,
-	}
-	r.Reset()
-	return r
+// storeRangeSet is an implementation of rangeSet which
+// cycles through a store's rangesByKey btree with mutex locking.
+type storeRangeSet struct {
+	store *Store
 }
 
-func (si *storeRangeIterator) Next() *Range {
-	si.store.mu.Lock()
-	defer si.store.mu.Unlock()
-	if index, remaining := si.index, len(si.store.rangesByKey)-si.index; remaining > 0 {
-		si.index++
-		si.remaining = remaining - 1
-		return si.store.rangesByKey[index]
-	}
-	return nil
-}
-
-func (si *storeRangeIterator) EstimatedCount() int {
-	return si.remaining
+func (rs *storeRangeSet) Visit(iterator func(*Range) bool) {
+	// Hold the lock except when the iterator is called.
```

Comment:
s/iterator/visitor/ (here and in the test)

Is this thread-safe? The docs don't discuss iterator invalidation that I can see, so it's hard to tell what's supposed to be guaranteed. It looks like you'll either process some nodes twice or skip some nodes if ranges are added or removed during the iteration, although I don't think it would panic (https://github.com/google/btree/blob/master/btree.go#L359). Not the end of the world here if we can get them on the next scan round, but again there is some risk of ranges being systematically skipped if we can't get through a scan cycle uninterrupted.